tpcc: add -partition flag

petermattis opened the pull request:
Partition all of the tables and indexes. Most of the tables are
partitioned by warehouse ID. The item table is partitioned by item
ID. The primary index of the history table is partitioned by
rowid (random 63-bit positive integer) while the other history table
indexes are partitioned by warehouse ID.

The motivation here is to be able to spread the partitions evenly in a
cluster in order to achieve better range distribution. This will require
setting zone configs for all of these partitions and starting up the
cluster with attributes or localities on each node.

petermattis committed:
tpcc: add -partition flag

Partition all of the tables and indexes. Most of the tables are
partitioned by warehouse ID. The item table is partitioned by item
ID. The primary index of the history table is partitioned by
rowid (random 63-bit positive integer) while the other history table
indexes are partitioned by warehouse ID.

The motivation here is to be able to spread the partitions evenly in a
cluster in order to achieve better range distribution. This will require
setting zone configs for all of these partitions and starting up the
cluster with attributes or localities on each node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/loadgen/156)
<!-- Reviewable:end -->

petermattis committed:
tpcc: specify zone configs for each partition

When using partitions, specify a zone config for each partition with a
required constraint.

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> cool! this is definitely the most interesting partitioning we've ever tested

there's a lot of repetition here and i didn't look closely for copy/paste errors but seems like it's doing the right thing

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/156#-:-L6r2FC046iDQdFWLGy1:bgg7ies)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[tpcc/partition.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/156#-L6r4RzSuenayKkSMo4z:-L6r4RzSuenayKkSMo5-:b-7f4akn) ([raw file](https://github.com/cockroachdb/loadgen/blob/f9a2ce94296b81811f058c2962688923de07a511/tpcc/partition.go#L198)):*
> ```Go
> }
> 
> func partitionItem(db *sql.DB, partitions int) {
> ```

The item table is read only. Maybe what we should do is make one copy of the item table per node, and change its zone config to have rf=1. Is that too hacky?

Haha just searched the spec for "replication" and found this gem:

```
Comment 2: If the implementation uses replication on the ITEM table and all transactions in Isolation Test 7 use the same copy of the ITEM table, updates to the ITEM table are not required to be propagated to other copies of the ITEM table. This relaxation of ACID properties on a replicated table is only valid under the above conditions and in the context of Isolation Test 7.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/156#-:-L6r56kOhTcv_zxFZKVa:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[tpcc/partition.go, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/loadgen/156#-L6r4RzSuenayKkSMo4z:-L6rJUunDaW7i8N04gCo:b-s8clme) ([raw file](https://github.com/cockroachdb/loadgen/blob/f9a2ce94296b81811f058c2962688923de07a511/tpcc/partition.go#L198)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

The item table is read only. Maybe what we should do is make one copy of the item table per node, and change its zone config to have rf=1. Is that too hacky?

Haha just searched the spec for "replication" and found this gem:

```
Comment 2: If the implementation uses replication on the ITEM table and all transactions in Isolation Test 7 use the same copy of the ITEM table, updates to the ITEM table are not required to be propagated to other copies of the ITEM table. This relaxation of ACID properties on a replicated table is only valid under the above conditions and in the context of Isolation Test 7.
```
</blockquote></details>

Yeah, I noticed that the `item` table is never modified. I don't think there is anything easy for us to do in the short term. When the max-safe-timestamp arrives, we can query at a historical timestamp which will allow reading from followers. And then we can increase the replication of the item table to provide more readable copies. Hopefully the work in 2.1 will allows this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/loadgen/156)*
<!-- Sent from Reviewable.io -->
